<a href="https://www.kaggle.com/code/emmanuelavogo/module-1-task?scriptVersionId=130974221" target="_blank"><img align="left" alt="Kaggle" title="Open in Kaggle" src="https://kaggle.com/static/images/open-in-kaggle.svg"></a>

# Module 1 Task

## Data Quality Assessment

Your task is to take a look at the following datasets provided by Sprocket Central Pty Ltd and identify all data quality issues. Once you've had a look at these datasets, draft an email to the client identifying all data quality issues.

Sprocket Central Pty Ltd provided KPMG with 4 datasets:

1. Transactions data
2. New Customer List
3. Customer Demographic
4. Customer Addresses

In [1]:
#Importing pandas library
import pandas as pd

In [2]:
## Reading the dataset
Dataset = pd.read_excel('/kaggle/input/kpmg-virtual-internship-project-task-1/KPMG Dataset.xlsx')


## Reading each sheet separately

### 1. Transactions

In [3]:
#reading the Transactions sheet
import pandas as pd
Transactions = pd.read_excel('/kaggle/input/kpmg-virtual-internship-project-task-1/KPMG Dataset.xlsx', sheet_name='Transactions')

In [4]:
#showing the first 5 rows
Transactions.head()

transaction_id  product_id  customer_id transaction_date  online_order  \
0               1           2         2950       2017-02-25           0.0   
1               2           3         3120       2017-05-21           1.0   
2               3          37          402       2017-10-16           0.0   
3               4          88         3135       2017-08-31           0.0   
4               5          78          787       2017-10-01           1.0   

  order_status           brand product_line product_class product_size  \
0     Approved           Solex     Standard        medium       medium   
1     Approved   Trek Bicycles     Standard        medium        large   
2     Approved      OHM Cycles     Standard           low       medium   
3     Approved  Norco Bicycles     Standard        medium       medium   
4     Approved  Giant Bicycles     Standard        medium        large   

   list_price  standard_cost product_first_sold_date  
0       71.49          53.62              2012-12-02  
1     2091.47         388.92              2014-03-03  
2     1793.43         248.82              1999-07-20  
3     1198.46         381.10              1998-12-16  
4     1765.30         709.48              2015-08-10

In [5]:
#displaying the column names and data types
Transactions.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20000 entries, 0 to 19999
Data columns (total 13 columns):
 #   Column                   Non-Null Count  Dtype         
---  ------                   --------------  -----         
 0   transaction_id           20000 non-null  int64         
 1   product_id               20000 non-null  int64         
 2   customer_id              20000 non-null  int64         
 3   transaction_date         20000 non-null  datetime64[ns]
 4   online_order             19640 non-null  float64       
 5   order_status             20000 non-null  object        
 6   brand                    19803 non-null  object        
 7   product_line             19803 non-null  object        
 8   product_class            19803 non-null  object        
 9   product_size             19803 non-null  object        
 10  list_price               20000 non-null  float64       
 11  standard_cost            19803 non-null  float64       
 12  product_first_sold_date  19803 n

In [6]:
#Counting the number of missing values
Transactions.isnull().sum()

transaction_id               0
product_id                   0
customer_id                  0
transaction_date             0
online_order               360
order_status                 0
brand                      197
product_line               197
product_class              197
product_size               197
list_price                   0
standard_cost              197
product_first_sold_date    197
dtype: int64

In [7]:
#dropping rows with missing values for online_order
Transactions.dropna(subset=['online_order'], inplace=True)

#dropping rows with missing values
Transactions.dropna(subset=['brand', 'product_line', 'product_class', 'product_size', 'standard_cost', 'product_first_sold_date'], inplace=True)

###### Dropping the rows where with null values because the columns have relatively small number of missing values (360 and 197) compared to the total number of rows in the data.

In [8]:
#Converting the 'product_first_sold_date' column to datetime format
Transactions['product_first_sold_date']=pd.to_datetime(Transactions[
    'product_first_sold_date'], format='%Y-%m-%d')

In [9]:
#Displaying basic statistics
Transactions.describe()

transaction_id    product_id   customer_id  online_order    list_price  \
count    19445.000000  19445.000000  19445.000000  19445.000000  19445.000000   
mean      9989.257393     45.797737   1739.467267      0.500849   1107.337193   
std       5779.669087     30.571996   1011.889153      0.500012    582.662400   
min          1.000000      0.000000      1.000000      0.000000     12.010000   
25%       4976.000000     18.000000    857.000000      0.000000    575.270000   
50%       9985.000000     45.000000   1741.000000      1.000000   1163.890000   
75%      14997.000000     72.000000   2615.000000      1.000000   1635.300000   
max      20000.000000    100.000000   5034.000000      1.000000   2091.470000   

       standard_cost  
count   19445.000000  
mean      555.585260  
std       405.596541  
min         7.210000  
25%       215.140000  
50%       507.580000  
75%       795.100000  
max      1759.850000

In [10]:
#checking for uniqueness of each column
Transactions.nunique()

transaction_id             19445
product_id                   101
customer_id                 3492
transaction_date             364
online_order                   2
order_status                   2
brand                          6
product_line                   4
product_class                  3
product_size                   3
list_price                   100
standard_cost                103
product_first_sold_date      100
dtype: int64

### Checking the columns

In [11]:
Transactions.columns

Index(['transaction_id', 'product_id', 'customer_id', 'transaction_date',
       'online_order', 'order_status', 'brand', 'product_line',
       'product_class', 'product_size', 'list_price', 'standard_cost',
       'product_first_sold_date'],
      dtype='object')

In [12]:
Transactions['order_status'].value_counts()

Approved     19273
Cancelled      172
Name: order_status, dtype: int64

In [13]:
Transactions['brand'].value_counts()

Solex             4169
WeareA2B          3245
Giant Bicycles    3244
OHM Cycles        2993
Trek Bicycles     2931
Norco Bicycles    2863
Name: brand, dtype: int64

In [14]:
Transactions['product_line'].value_counts()

Standard    13920
Road         3894
Touring      1213
Mountain      418
Name: product_line, dtype: int64

In [15]:
Transactions['product_class'].value_counts()

medium    13587
high       2952
low        2906
Name: product_class, dtype: int64

In [16]:
Transactions['product_size'].value_counts()

medium    12767
large      3900
small      2778
Name: product_size, dtype: int64

In [17]:
Transactions['product_first_sold_date'].head(20)

0    2012-12-02
1    2014-03-03
2    1999-07-20
3    1998-12-16
4    2015-08-10
5    2006-11-10
6    1993-07-15
7    2009-04-12
8    1991-08-05
9    2004-08-17
10   2011-08-29
11   2015-06-17
12   1994-07-12
13   1994-09-09
14   2004-07-25
15   2003-09-09
16   2004-08-07
17   1992-10-11
18   2009-04-12
19   2002-03-22
Name: product_first_sold_date, dtype: datetime64[ns]

### 2. New Customer List

In [18]:
#Reading the NewCustomerList sheet
New_Customer_List = pd.read_excel('/kaggle/input/kpmg-virtual-internship-project-task-1/KPMG Dataset.xlsx', sheet_name="NewCustomerList")

/opt/conda/lib/python3.7/site-packages/ipykernel_launcher.py:2: FutureWarning: Inferring datetime64[ns] from data containing strings is deprecated and will be removed in a future version. To retain the old behavior explicitly pass Series(data, dtype={value.dtype})
  


In [19]:
#specifying the data type for the column using the dtype to avoid the warning
New_Customer_List = pd.read_excel('/kaggle/input/kpmg-virtual-internship-project-task-1/KPMG Dataset.xlsx', sheet_name='NewCustomerList', dtype={'DOB': 'datetime64[ns]'})

In [20]:
#showing the first 5 rows
New_Customer_List.head()

first_name  last_name  gender  past_3_years_bike_related_purchases  \
0    Chickie    Brister    Male                                   86   
1      Morly     Genery    Male                                   69   
2    Ardelis  Forrester  Female                                   10   
3     Lucine      Stutt  Female                                   64   
4    Melinda     Hadlee  Female                                   34   

         DOB                   job_title job_industry_category  \
0 1957-07-12             General Manager         Manufacturing   
1 1970-03-22         Structural Engineer              Property   
2 1974-08-28      Senior Cost Accountant    Financial Services   
3 1979-01-28  Account Representative III         Manufacturing   
4 1965-09-21           Financial Analyst    Financial Services   

      wealth_segment deceased_indicator owns_car  ...  state    country  \
0      Mass Customer                  N      Yes  ...    QLD  Australia   
1      Mass Customer                  N       No  ...    NSW  Australia   
2  Affluent Customer                  N       No  ...    VIC  Australia   
3  Affluent Customer                  N      Yes  ...    QLD  Australia   
4  Affluent Customer                  N       No  ...    NSW  Australia   

   property_valuation Unnamed: 16 Unnamed: 17  Unnamed: 18  Unnamed: 19  \
0                   6        0.92      1.1500       1.4375     1.221875   
1                  11        0.57      0.5700       0.7125     0.605625   
2                   5        0.82      0.8200       0.8200     0.820000   
3                   1        1.05      1.3125       1.3125     1.312500   
4                   9        0.78      0.7800       0.9750     0.975000   

   Unnamed: 20  Rank     Value  
0            1     1  1.718750  
1            1     1  1.718750  
2            1     1  1.718750  
3            4     4  1.703125  
4            4     4  1.703125  

[5 rows x 23 columns]

In [21]:
#displaying the column names and data types
New_Customer_List.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 23 columns):
 #   Column                               Non-Null Count  Dtype         
---  ------                               --------------  -----         
 0   first_name                           1000 non-null   object        
 1   last_name                            971 non-null    object        
 2   gender                               1000 non-null   object        
 3   past_3_years_bike_related_purchases  1000 non-null   int64         
 4   DOB                                  983 non-null    datetime64[ns]
 5   job_title                            894 non-null    object        
 6   job_industry_category                835 non-null    object        
 7   wealth_segment                       1000 non-null   object        
 8   deceased_indicator                   1000 non-null   object        
 9   owns_car                             1000 non-null   object        
 10  tenure       

In [22]:
#Dropping the unnamed columns
New_Customer_List.drop(['Unnamed: 16', 'Unnamed: 17', 'Unnamed: 18',
 'Unnamed: 19', 'Unnamed: 20'], axis=1, inplace=True)

In [23]:
#Counting the number of missing values
New_Customer_List.isnull().sum()

first_name                               0
last_name                               29
gender                                   0
past_3_years_bike_related_purchases      0
DOB                                     17
job_title                              106
job_industry_category                  165
wealth_segment                           0
deceased_indicator                       0
owns_car                                 0
tenure                                   0
address                                  0
postcode                                 0
state                                    0
country                                  0
property_valuation                       0
Rank                                     0
Value                                    0
dtype: int64

In [24]:
#Counting the number of duplicate rows
New_Customer_List.duplicated().sum()

0

In [25]:
#dropping rows with missing values for first_name and last_name
New_Customer_List.dropna(subset=['last_name', 'DOB'], inplace=True)

In [26]:
#replacing missing values
New_Customer_List['job_title'].fillna(value='Unknown', inplace=True)
New_Customer_List['job_industry_category'].fillna(value='Unknown', inplace=True)

In [27]:
#displaying basic statistics
New_Customer_List.describe()

past_3_years_bike_related_purchases      tenure     postcode  \
count                           954.000000  954.000000   954.000000   
mean                             49.958071   11.446541  3016.632075   
std                              27.868382    5.019978   849.825067   
min                               0.000000    1.000000  2000.000000   
25%                              27.000000    8.000000  2209.000000   
50%                              51.000000   11.000000  2782.500000   
75%                              72.000000   15.000000  3850.750000   
max                              99.000000   22.000000  4879.000000   

       property_valuation         Rank       Value  
count          954.000000   954.000000  954.000000  
mean             7.422432   497.243187    0.883398  
std              2.741152   290.127232    0.294979  
min              1.000000     1.000000    0.340000  
25%              6.000000   244.000000    0.654937  
50%              8.000000   501.000000    0.859250  
75%              9.000000   746.250000    1.083750  
max             12.000000  1000.000000    1.718750

### Checking the columns

In [28]:
#Counting the number of occurrences of each column
New_Customer_List.columns

Index(['first_name', 'last_name', 'gender',
       'past_3_years_bike_related_purchases', 'DOB', 'job_title',
       'job_industry_category', 'wealth_segment', 'deceased_indicator',
       'owns_car', 'tenure', 'address', 'postcode', 'state', 'country',
       'property_valuation', 'Rank', 'Value'],
      dtype='object')

In [29]:
New_Customer_List['gender'].value_counts()

Female    499
Male      455
Name: gender, dtype: int64

In [30]:
New_Customer_List['DOB'].value_counts()

1994-04-15    2
1997-10-23    2
1963-08-25    2
1978-01-15    2
1992-10-13    2
             ..
1997-09-27    1
1978-05-27    1
1945-08-08    1
1943-08-27    1
1955-10-02    1
Name: DOB, Length: 932, dtype: int64

In [31]:
New_Customer_List['job_industry_category'].value_counts()

Manufacturing         195
Financial Services    194
Unknown               159
Health                149
Retail                 78
Property               62
IT                     35
Entertainment          34
Argiculture            25
Telecommunications     23
Name: job_industry_category, dtype: int64

In [32]:
New_Customer_List['wealth_segment'].value_counts()

Mass Customer        488
High Net Worth       241
Affluent Customer    225
Name: wealth_segment, dtype: int64

In [33]:
New_Customer_List['state'].value_counts()

NSW    485
VIC    250
QLD    219
Name: state, dtype: int64

In [34]:
New_Customer_List['owns_car'].value_counts()

No     482
Yes    472
Name: owns_car, dtype: int64

In [35]:
New_Customer_List['deceased_indicator'].value_counts()

N    954
Name: deceased_indicator, dtype: int64

### 3. Customer Demographic

In [36]:
#Reading the CustomerDemographic sheet
Customer_Demographic = pd.read_excel('/kaggle/input/kpmg-virtual-internship-project-task-1/KPMG Dataset.xlsx', 
                                     sheet_name='CustomerDemographic', 
                                     dtype={'DOB': 'datetime64[ns]'})

In [37]:
#showing the first 5 rows
Customer_Demographic.head()

customer_id      first_name  last_name  gender  \
0            1         Laraine  Medendorp       F   
1            2             Eli    Bockman    Male   
2            3           Arlin     Dearle    Male   
3            4          Talbot        NaN    Male   
4            5  Sheila-kathryn     Calton  Female   

   past_3_years_bike_related_purchases        DOB               job_title  \
0                                   93 1953-10-12     Executive Secretary   
1                                   81 1980-12-16  Administrative Officer   
2                                   61 1954-01-20      Recruiting Manager   
3                                   33 1961-10-03                     NaN   
4                                   56 1977-05-13           Senior Editor   

  job_industry_category     wealth_segment deceased_indicator  \
0                Health      Mass Customer                  N   
1    Financial Services      Mass Customer                  N   
2              Property      Mass Customer                  N   
3                    IT      Mass Customer                  N   
4                   NaN  Affluent Customer                  N   

                                             default owns_car  tenure  
0                                                 "'      Yes    11.0  
1                       <script>alert('hi')</script>      Yes    16.0  
2                                2018-02-01 00:00:00      Yes    15.0  
3  () { _; } >_[$($())] { touch /tmp/blns.shellsh...       No     7.0  
4                                                NIL      Yes     8.0

In [38]:
#displaying the column names and data types
Customer_Demographic.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4000 entries, 0 to 3999
Data columns (total 13 columns):
 #   Column                               Non-Null Count  Dtype         
---  ------                               --------------  -----         
 0   customer_id                          4000 non-null   int64         
 1   first_name                           4000 non-null   object        
 2   last_name                            3875 non-null   object        
 3   gender                               4000 non-null   object        
 4   past_3_years_bike_related_purchases  4000 non-null   int64         
 5   DOB                                  3913 non-null   datetime64[ns]
 6   job_title                            3494 non-null   object        
 7   job_industry_category                3344 non-null   object        
 8   wealth_segment                       4000 non-null   object        
 9   deceased_indicator                   4000 non-null   object        
 10  default     

In [39]:
#Counting the number of missing values
Customer_Demographic.isnull().sum()

customer_id                              0
first_name                               0
last_name                              125
gender                                   0
past_3_years_bike_related_purchases      0
DOB                                     87
job_title                              506
job_industry_category                  656
wealth_segment                           0
deceased_indicator                       0
default                                302
owns_car                                 0
tenure                                  87
dtype: int64

In [40]:
#Counting the number of duplicate rows
Customer_Demographic.duplicated().sum()

0

In [41]:
#Dropping rows with missing values for first_name and last_name
Customer_Demographic.dropna(subset=['last_name', 'DOB', 'tenure'], 
                            inplace=True)

In [42]:
#Replacing missing values
Customer_Demographic['job_title'].fillna(value='Unknown', inplace=True)
Customer_Demographic['job_industry_category'].fillna(value='Unknown', 
                                                     inplace=True)

In [43]:
#Replacing gender abbreviations with full names
Customer_Demographic['gender'].replace(['Femal', 'F'], 'Female', inplace=True)
Customer_Demographic['gender'].replace('M', 'Male', inplace=True)
Customer_Demographic['gender'].replace('U', 'Unknown', inplace=True)

In [44]:
#Converting tenure to int
Customer_Demographic['tenure'] = Customer_Demographic['tenure'].astype(int)

In [45]:
#displaying basic statistics for numerical columns
Customer_Demographic.describe()

customer_id  past_3_years_bike_related_purchases       tenure
count  3792.000000                          3792.000000  3792.000000
mean   1996.283228                            48.983122    10.674578
std    1155.001491                            28.755975     5.661852
min       1.000000                             0.000000     1.000000
25%    1000.750000                            24.000000     6.000000
50%    1994.500000                            48.000000    11.000000
75%    2991.250000                            73.000000    15.000000
max    4000.000000                            99.000000    22.000000

### Checking the columns

In [46]:
Customer_Demographic['gender'].value_counts()

Female     1982
Male       1809
Unknown       1
Name: gender, dtype: int64

In [47]:
Customer_Demographic['past_3_years_bike_related_purchases'].value_counts()

16    55
19    54
20    53
67    49
2     48
      ..
8     27
58    26
86    26
95    25
92    22
Name: past_3_years_bike_related_purchases, Length: 100, dtype: int64

In [48]:
Customer_Demographic['DOB'].value_counts()

1978-01-30    7
1976-09-25    4
1977-05-13    4
1978-08-19    4
1964-07-08    4
             ..
1978-06-18    1
1989-06-16    1
1998-09-30    1
1985-03-11    1
1991-11-05    1
Name: DOB, Length: 3351, dtype: int64

In [49]:
Customer_Demographic['job_title'].value_counts()

Unknown                                 481
Business Systems Development Analyst     42
Social Worker                            42
Tax Accountant                           42
Recruiting Manager                       41
                                       ... 
Research Assistant III                    3
Health Coach III                          3
Health Coach I                            3
Database Administrator I                  3
Developer I                               1
Name: job_title, Length: 196, dtype: int64

In [50]:
Customer_Demographic['job_industry_category'].value_counts()

Manufacturing         768
Financial Services    742
Unknown               631
Health                582
Retail                348
Property              264
IT                    148
Entertainment         130
Argiculture           111
Telecommunications     68
Name: job_industry_category, dtype: int64

In [51]:
Customer_Demographic['wealth_segment'].value_counts()

Mass Customer        1901
High Net Worth        955
Affluent Customer     936
Name: wealth_segment, dtype: int64

In [52]:
Customer_Demographic['deceased_indicator'].value_counts()

N    3790
Y       2
Name: deceased_indicator, dtype: int64

In [53]:
Customer_Demographic['default'].value_counts()

1                                         110
100                                       108
-1                                        104
-100                                       95
â°â´âµâââ                                  52
                                         ... 
testâ testâ«                               30
âªâªtest⪠                                29
×Ö¸×ְתָ×testاÙصÙحات اÙتÙØ­ÙÙ     29
,ãã»:*:ã»ãâ( â» Ï â» )ãã»:*:ã»ãâ           24
ì¸ëë°í ë¥´                                 24
Name: default, Length: 90, dtype: int64

In [54]:
#dropping the column since the values are inconsistent
Customer_Demographic = Customer_Demographic.drop('default', axis=1)

In [55]:
Customer_Demographic['owns_car'].value_counts()

Yes    1917
No     1875
Name: owns_car, dtype: int64

### Note that we first dropped the rows with missing values for "last_name" and "DOB" in both the "NewCustomerList" and "CustomerDemographic" dataframes, as we would be joining the dataframes using these columns.

### 4.  Customer Address

In [56]:
#reading the Transactions sheet from the dataset and store it as a dataframe
Customer_Address = pd.read_excel('/kaggle/input/kpmg-virtual-internship-project-task-1/KPMG Dataset.xlsx', 
                                 sheet_name='CustomerAddress')

In [57]:
#showing the first 5 rows
Customer_Address.head()

customer_id              address  postcode            state    country  \
0            1   060 Morning Avenue      2016  New South Wales  Australia   
1            2  6 Meadow Vale Court      2153  New South Wales  Australia   
2            4   0 Holy Cross Court      4211              QLD  Australia   
3            5  17979 Del Mar Point      2448  New South Wales  Australia   
4            6     9 Oakridge Court      3216              VIC  Australia   

   property_valuation  
0                  10  
1                  10  
2                   9  
3                   4  
4                   9

In [58]:
#displaying the column names and data types
Customer_Address.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3999 entries, 0 to 3998
Data columns (total 6 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   customer_id         3999 non-null   int64 
 1   address             3999 non-null   object
 2   postcode            3999 non-null   int64 
 3   state               3999 non-null   object
 4   country             3999 non-null   object
 5   property_valuation  3999 non-null   int64 
dtypes: int64(3), object(3)
memory usage: 187.6+ KB


In [59]:
#Checking for null values
Customer_Address.isnull().sum()

customer_id           0
address               0
postcode              0
state                 0
country               0
property_valuation    0
dtype: int64

In [60]:
#Checking for duplicate values
Customer_Address.duplicated().sum()

0

In [61]:
#Checking for uniqueness of each column
Customer_Address.nunique()

customer_id           3999
address               3996
postcode               873
state                    5
country                  1
property_valuation      12
dtype: int64

### Checking the columns

In [62]:
Customer_Address['postcode'].value_counts()

2170    31
2155    30
2145    30
2153    29
3977    26
        ..
3808     1
3114     1
4721     1
4799     1
3089     1
Name: postcode, Length: 873, dtype: int64

In [63]:
Customer_Address['state'].value_counts()

NSW                2054
VIC                 939
QLD                 838
New South Wales      86
Victoria             82
Name: state, dtype: int64

In [64]:
Customer_Address['country'].value_counts()

Australia    3999
Name: country, dtype: int64

In [65]:
Customer_Address['property_valuation'].value_counts()

9     647
8     646
10    577
7     493
11    281
6     238
5     225
4     214
12    195
3     186
1     154
2     143
Name: property_valuation, dtype: int64

### Email to client

Subject: Data Quality Issues Identified in the Sprocket Central Pty Ltd Dataset

Hello,

I am writing this email to identify some data quality issues with the four datasets that you have provided to us, namely Transactions, New Customer List, Customer Demographic, and Customer Addresses.

1. Transactions:

There are missing values in the 'online_order', 'brand', 'product_line', 'product_class', 'product_size', 'standard_cost', and 'product_first_sold_date' columns. We have dropped rows with missing values for 'online_order' and the above columns.
There are some duplicate 'transaction_id' values.
The values in the 'product_first_sold_date' column were not in datetime format, and we have converted them accordingly.

2. New Customer List:

There are missing values in the 'last_name' and 'DOB' columns, which we have dropped.
There are some duplicate 'customer_id' values.
There are missing values in the 'job_title' and 'job_industry_category' columns, which we have replaced.
There are unnamed columns that we have dropped.

3. Customer Demographic:

There are missing values in the 'last_name', 'DOB', and 'tenure' columns, which we have dropped.
There are some duplicate 'customer_id' values.
There are missing values in the 'job_title' and 'job_industry_category' columns, which we have replaced.
The values in the 'gender' column have inconsistent values, which we have replaced.
The 'default' column has inconsistent values, so we have dropped it.

4. Customer Addresses:

The data is good for analysis.

We recommend that you review and clean the data to improve the accuracy of the analysis. Please let us know if you have any questions or concerns.

Thank you for your attention.

Best regards,

Emmanuel Avogo.